### PR TITLE
[asana] Multiple fixes: string GIDs, Client as constructor, FindAllParams interfaces, is_organization

### PR DESCRIPTION
**Pull request by giact**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/Asana/node-asana/ and https://developers.asana.com/docs/
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

***

* I fixed multiple issues that I encountered when using these type definitions for my project.
* I added all the relevant test cases that I could think of, using as inspiration both my project and the [`node-asana` tests](https://github.com/Asana/node-asana/tree/master/test/resources)
* Even if my project uses  the latest `node-asana` `0.18.5`, I do not think I changed anything that should not have already be present in `0.18.4`, so I did not change the version here
* During my tests, I noticed that the type for `client.tasks.setParent()` is missing, but I am not experienced enough with these projects to add it, so I mentioned it as a commented out **TODO** line in the tests

### CODE REVIEW

File: types/asana/index.d.ts [hunk 1]
```diff
@@ -1480,7 +1484,7 @@ declare namespace asana {
              * * you last retrieved the task.
              * *
              * * Returns the complete updated tag record.
-             *   * @param {Number} tag The tag to update.
+             *   * @param {String} tag The tag to update.
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:1487]:
```suggestion
             *   * @param {string} tag The tag to update.
```

File: types/asana/index.d.ts [hunk 2]
```diff
@@ -1489,21 +1493,21 @@ declare namespace asana {
...
-            update(tag: number, data: Tags.Type, dispatchOptions?: any): Promise<Tags.Type>;
+            update(tag: string | number, data: Tags.Type, dispatchOptions?: any): Promise<Tags.Type>;
 
             /**
              * * A specific, existing tag can be deleted by making a DELETE request
              * * on the URL for that tag.
              * *
              * * Returns an empty data record.
-             *   * @param {Number} tag The tag to delete.
+             *   * @param {String} tag The tag to delete.
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:1503]:
```suggestion
             *   * @param {string} tag The tag to delete.
```

File: types/asana/index.d.ts [hunk 3]
```diff
@@ -1523,7 +1527,7 @@ declare namespace asana {
 
             /**
              * * Returns the compact tag records for all tags in the workspace.
-             *   * @param {Number} workspace The workspace or organization to find tags in.
+             *   * @param {String} workspace The workspace or organization to find tags in.
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:1530]:
```suggestion
             *   * @param {string} workspace The workspace or organization to find tags in.
```

File: types/asana/index.d.ts [hunk 4]
```diff
@@ -1533,15 +1537,15 @@ declare namespace asana {
...
+                workspace: string | number,
                 params?: PaginationParams,
                 dispatchOptions?: any,
             ): Promise<ResourceList<Tags.Type>>;
 
             /**
              * * Returns the compact task records for all tasks with the given tag.
              * * Tasks can have more than one tag at a time.
-             *   * @param {Number} tag The tag to fetch tasks from.
+             *   * @param {String} tag The tag to fetch tasks from.
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:1548]:
```suggestion
             *   * @param {string} tag The tag to fetch tasks from.
```

File: types/asana/index.d.ts [hunk 5]
```diff
@@ -1663,7 +1669,7 @@ declare namespace asana {
              * * Every task is required to be created in a specific workspace, and this
              * * workspace cannot be changed once set. The workspace need not be set
              * * explicitly if you specify a `project` or a `parent` task instead.
-             *   * @param {Number} workspace The workspace to create a task in.
+             *   * @param {String} workspace The workspace to create a task in.
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:1672]:
```suggestion
             *   * @param {string} workspace The workspace to create a task in.
```

File: types/asana/index.d.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ declare namespace asana {
          * @param {String} [redirectUri]  Default redirect URI for this client
          * @param {String} [asanaBaseUrl] Base URL for Asana, for debugging
          */
-        (dispatcher: Dispatcher, options?: ClientOptions): asana.Client;
+        new (dispatcher: Dispatcher, options?: ClientOptions): asana.Client;
```
Comment (peterblazejewicz) [inline types/asana/index.d.ts:28]:
```suggestion
        new (dispatcher: Dispatcher, options?: ClientOptions): Client;
```

**Comment by peterblazejewicz** [comment]
@giact 
so those identifiers with type like `string | number` (`attachment`, etc) are typed as `string` in JSDoc now, though the params are cleary stating `string | number`.
that is perfectly OK to have multiple types in JSDoc param type, if you intend to use types in JSDoc:
```
* @param {string|number} [someOption] - description
```
https://jsdoc.app/tags-param.html#multiple-types-and-repeatable-parameters
I can see library itself is not consistent on documenting the param type (the JSDOc mentionds strings, while all tests seem to use numbers)

**Comment by giact** [comment]
@peterblazejewicz
yes, I was unsure how to proceed with the JSDoc because, as you said, the library states they are string while their test code still uses both.

I assumed that Asana wants everyone to use string and deprecate (and actively discourage) the use of numbers because of the 2^53 issue as explained here:
https://forum.asana.com/t/asana-is-moving-to-string-ids-updated-with-revised-timeline/29340
While they simply left the old tests as is for legacy reasons.

If you think it's better to change the JSDoc as `{String|Number}` I'll do so, and maybe add in each description a warning about not using number (what do you think?)

**Comment by peterblazejewicz** [comment]
They said:
> Additionally, places where the API accepts IDs will begin to accept both numbers and strings.

So let's do exactly this in JSDoc param type (if you still stick to JSDoc param, as in TS that's not required).
I don't think you should enforce one type other another. In the handler for the input they are using methods (format) from util, which just works with any type IMO

**Comment by giact** [comment]
I changed all the `@param` that refer to GIDs that I could find, including some that were not touched by my fix.
I kept the same style (with uppercase: `{String|Number}`) that is already in use elsewhere in these JSDocs

**Review by peterblazejewicz — APPROVED** [review]
LGMT!
@giact thx!
let's wait for owners review

**Comment by filipposarzana** [comment]
Ready to merge

**Review by filipposarzana — APPROVED** [review]

**Pull request merged**